<!DOCTYPE html>

<html :class="{'dark': darkMode === 'dark' || (darkMode === 'system' &amp;&amp; window.matchMedia('(prefers-color-scheme: dark)').matches)}" class="scroll-smooth" data-content_root="./" lang="en" x-data="{ darkMode: localStorage.getItem('darkMode') || localStorage.setItem('darkMode', 'system'), activeSection: '' }" x-init="$watch('darkMode', val =&gt; localStorage.setItem('darkMode', val))">
<head>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<meta charset="utf-8"/>
<meta content="white" media="(prefers-color-scheme: light)" name="theme-color"/>
<meta content="black" media="(prefers-color-scheme: dark)" name="theme-color"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<title>Complete Documentation | H2M 1.0.0 documentation</title>
<meta content="Complete Documentation | H2M 1.0.0 documentation" property="og:title"/>
<meta content="Complete Documentation | H2M 1.0.0 documentation" name="twitter:title"/>
<link href="_static/pygments.css?v=8d216cef" rel="stylesheet" type="text/css"/>
<link href="_static/theme.css?v=8958d2b4" rel="stylesheet" type="text/css"/>
<link href="_static/css/custom.css" rel="stylesheet" type="text/css"/>
<link href="search.html" rel="search" title="Search"/>
<link href="genindex.html" rel="index" title="Index"/>
<link href="jupyter.html" rel="next" title="Jupyter Notebook Tutorial"/>
<link href="quickstart.html" rel="prev" title="Quick Start"/>
<script>
    <!-- Prevent Flash of wrong theme -->
      const userPreference = localStorage.getItem('darkMode');
      let mode;
      if (userPreference === 'dark' || window.matchMedia('(prefers-color-scheme: dark)').matches) {
        mode = 'dark';
        document.documentElement.classList.add('dark');
      } else {
        mode = 'light';
      }
      if (!userPreference) {localStorage.setItem('darkMode', mode)}
    </script>
</head>
<body :class="{ 'overflow-hidden': showSidebar }" class="min-h-screen font-sans antialiased bg-background text-foreground" x-data="{ showSidebar: false }">
<div @click.self="showSidebar = false" class="fixed inset-0 z-50 overflow-hidden bg-background/80 backdrop-blur-sm md:hidden" x-cloak="" x-show="showSidebar"></div><div class="relative flex flex-col min-h-screen" id="page"><a class="absolute top-0 left-0 z-[100] block bg-background p-4 text-xl transition -translate-x-full opacity-0 focus:translate-x-0 focus:opacity-100" href="#content">
      Skip to content
    </a><header class="sticky top-0 z-40 w-full border-b shadow-sm border-border supports-backdrop-blur:bg-background/60 bg-background/95 backdrop-blur"><div class="container flex items-center h-14">
<div class="hidden mr-4 md:flex">
<a class="flex items-center mr-6" href="index.html">
<img alt="Logo" class="mr-2 hidden dark:block" height="24" src="_static/h2m-logo-final.png" width="24"/>
<img alt="Logo" class="mr-2 dark:hidden" height="24" src="_static/h2m-logo-final.png" width="24"/><span class="hidden font-bold sm:inline-block text-clip whitespace-nowrap">H2M 1.0.0 documentation</span>
</a></div><button @click="showSidebar = true" class="inline-flex items-center justify-center h-10 px-0 py-2 mr-2 text-base font-medium transition-colors rounded-md hover:text-accent-foreground hover:bg-transparent md:hidden" type="button">
<svg aria-hidden="true" fill="currentColor" height="24" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M152.587 825.087q-19.152 0-32.326-13.174t-13.174-32.326q0-19.152 13.174-32.326t32.326-13.174h440q19.152 0 32.326 13.174t13.174 32.326q0 19.152-13.174 32.326t-32.326 13.174h-440Zm0-203.587q-19.152 0-32.326-13.174T107.087 576q0-19.152 13.174-32.326t32.326-13.174h320q19.152 0 32.326 13.174T518.087 576q0 19.152-13.174 32.326T472.587 621.5h-320Zm0-203.587q-19.152 0-32.326-13.174t-13.174-32.326q0-19.152 13.174-32.326t32.326-13.174h440q19.152 0 32.326 13.174t13.174 32.326q0 19.152-13.174 32.326t-32.326 13.174h-440ZM708.913 576l112.174 112.174q12.674 12.674 12.674 31.826t-12.674 31.826Q808.413 764.5 789.261 764.5t-31.826-12.674l-144-144Q600 594.391 600 576t13.435-31.826l144-144q12.674-12.674 31.826-12.674t31.826 12.674q12.674 12.674 12.674 31.826t-12.674 31.826L708.913 576Z"></path>
</svg>
<span class="sr-only">Toggle navigation menu</span>
</button>
<div class="flex items-center justify-between flex-1 space-x-2 sm:space-x-4 md:justify-end">
<div class="flex-1 w-full md:w-auto md:flex-none"><form @keydown.k.window.meta="$refs.search.focus()" action="search.html" class="relative flex items-center group" id="searchbox" method="get">
<input aria-label="Search the docs" class="inline-flex items-center font-medium transition-colors bg-transparent focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 ring-offset-background border border-input hover:bg-accent focus:bg-accent hover:text-accent-foreground focus:text-accent-foreground hover:placeholder-accent-foreground py-2 px-4 relative h-9 w-full justify-start rounded-[0.5rem] text-sm text-muted-foreground sm:pr-12 md:w-40 lg:w-64" id="search-input" name="q" placeholder="Search ..." type="search" x-ref="search"/>
<kbd class="pointer-events-none absolute right-1.5 top-2 hidden h-5 select-none text-muted-foreground items-center gap-1 rounded border border-border bg-muted px-1.5 font-mono text-[10px] font-medium opacity-100 sm:flex group-hover:bg-accent group-hover:text-accent-foreground">
<span class="text-xs">⌘</span>
    K
  </kbd>
</form>
</div>
<nav class="flex items-center space-x-1">
<button @click="darkMode = darkMode === 'light' ? 'dark' : 'light'" aria-label="Color theme switcher" class="relative inline-flex items-center justify-center px-0 text-sm font-medium transition-colors rounded-md hover:bg-accent hover:text-accent-foreground h-9 w-9" type="button">
<svg class="absolute transition-all scale-100 rotate-0 dark:-rotate-90 dark:scale-0" fill="currentColor" height="24" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M480 685q45.456 0 77.228-31.772Q589 621.456 589 576q0-45.456-31.772-77.228Q525.456 467 480 467q-45.456 0-77.228 31.772Q371 530.544 371 576q0 45.456 31.772 77.228Q434.544 685 480 685Zm0 91q-83 0-141.5-58.5T280 576q0-83 58.5-141.5T480 376q83 0 141.5 58.5T680 576q0 83-58.5 141.5T480 776ZM80 621.5q-19.152 0-32.326-13.174T34.5 576q0-19.152 13.174-32.326T80 530.5h80q19.152 0 32.326 13.174T205.5 576q0 19.152-13.174 32.326T160 621.5H80Zm720 0q-19.152 0-32.326-13.174T754.5 576q0-19.152 13.174-32.326T800 530.5h80q19.152 0 32.326 13.174T925.5 576q0 19.152-13.174 32.326T880 621.5h-80Zm-320-320q-19.152 0-32.326-13.174T434.5 256v-80q0-19.152 13.174-32.326T480 130.5q19.152 0 32.326 13.174T525.5 176v80q0 19.152-13.174 32.326T480 301.5Zm0 720q-19.152 0-32.326-13.17Q434.5 995.152 434.5 976v-80q0-19.152 13.174-32.326T480 850.5q19.152 0 32.326 13.174T525.5 896v80q0 19.152-13.174 32.33-13.174 13.17-32.326 13.17ZM222.174 382.065l-43-42Q165.5 327.391 166 308.239t13.174-33.065q13.435-13.674 32.587-13.674t32.065 13.674l42.239 43q12.674 13.435 12.555 31.706-.12 18.272-12.555 31.946-12.674 13.674-31.445 13.413-18.772-.261-32.446-13.174Zm494 494.761-42.239-43q-12.674-13.435-12.674-32.087t12.674-31.565Q686.609 756.5 705.38 757q18.772.5 32.446 13.174l43 41.761Q794.5 824.609 794 843.761t-13.174 33.065Q767.391 890.5 748.239 890.5t-32.065-13.674Zm-42-494.761Q660.5 369.391 661 350.62q.5-18.772 13.174-32.446l41.761-43Q728.609 261.5 747.761 262t33.065 13.174q13.674 13.435 13.674 32.587t-13.674 32.065l-43 42.239q-13.435 12.674-31.706 12.555-18.272-.12-31.946-12.555Zm-495 494.761Q165.5 863.391 165.5 844.239t13.674-32.065l43-42.239q13.435-12.674 32.087-12.674t31.565 12.674Q299.5 782.609 299 801.38q-.5 18.772-13.174 32.446l-41.761 43Q231.391 890.5 212.239 890t-33.065-13.174ZM480 576Z"></path>
</svg>
<svg class="absolute transition-all scale-0 rotate-90 dark:rotate-0 dark:scale-100" fill="currentColor" height="24" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M480 936q-151 0-255.5-104.5T120 576q0-138 90-239.5T440 218q25-3 39 18t-1 44q-17 26-25.5 55t-8.5 61q0 90 63 153t153 63q31 0 61.5-9t54.5-25q21-14 43-1.5t19 39.5q-14 138-117.5 229T480 936Zm0-80q88 0 158-48.5T740 681q-20 5-40 8t-40 3q-123 0-209.5-86.5T364 396q0-20 3-40t8-40q-78 32-126.5 102T200 576q0 116 82 198t198 82Zm-10-270Z"></path>
</svg>
</button>
</nav>
</div>
</div>
</header>
<div class="flex-1"><div class="container flex-1 items-start md:grid md:grid-cols-[220px_minmax(0,1fr)] md:gap-6 lg:grid-cols-[240px_minmax(0,1fr)] lg:gap-10"><aside :aria-hidden="!showSidebar" :class="{ 'translate-x-0': showSidebar }" class="fixed inset-y-0 left-0 md:top-14 z-50 md:z-30 bg-background md:bg-transparent transition-all duration-100 -translate-x-full md:translate-x-0 ml-0 p-6 md:p-0 md:-ml-2 md:h-[calc(100vh-3.5rem)] w-5/6 md:w-full shrink-0 overflow-y-auto border-r border-border md:sticky" id="left-sidebar">
<a class="!justify-start text-sm md:!hidden bg-background" href="index.html">
<img alt="Logo" class="mr-2 hidden dark:block" height="16" src="_static/h2m-logo-final.png" width="16"/>
<img alt="Logo" class="mr-2 dark:hidden" height="16" src="_static/h2m-logo-final.png" width="16"/><span class="font-bold text-clip whitespace-nowrap">H2M 1.0.0 documentation</span>
</a>
<div class="relative overflow-hidden md:overflow-auto my-4 md:my-0 h-[calc(100vh-8rem)] md:h-auto">
<div class="overflow-y-auto h-full w-full relative pr-6"><nav class="table w-full min-w-full my-6 lg:my-8">
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Complete Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="jupyter.html">Jupyter Notebook Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream.html">COMING SOON - H2M Database Online Portal</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About H2M</a></li>
</ul>
</nav>
</div>
</div>
<button @click="showSidebar = false" class="absolute md:hidden right-4 top-4 rounded-sm opacity-70 transition-opacity hover:opacity-100" type="button">
<svg class="h-4 w-4" fill="currentColor" height="24" stroke="none" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M480 632 284 828q-11 11-28 11t-28-11q-11-11-11-28t11-28l196-196-196-196q-11-11-11-28t11-28q11-11 28-11t28 11l196 196 196-196q11-11 28-11t28 11q11 11 11 28t-11 28L536 576l196 196q11 11 11 28t-11 28q-11 11-28 11t-28-11L480 632Z"></path>
</svg>
</button>
</aside>
<main class="relative py-6 lg:gap-10 lg:py-8 xl:grid xl:grid-cols-[1fr_300px]">
<div class="w-full min-w-0 mx-auto">
<nav aria-label="breadcrumbs" class="flex items-center mb-4 space-x-1 text-sm text-muted-foreground">
<a class="overflow-hidden text-ellipsis whitespace-nowrap hover:text-foreground" href="index.html">
<span class="hidden md:inline">H2M 1.0.0 documentation</span>
<svg aria-label="Home" class="md:hidden" fill="currentColor" height="18" stroke="none" viewbox="0 96 960 960" width="18" xmlns="http://www.w3.org/2000/svg">
<path d="M240 856h120V616h240v240h120V496L480 316 240 496v360Zm-80 80V456l320-240 320 240v480H520V696h-80v240H160Zm320-350Z"></path>
</svg>
</a>
<div class="mr-1">/</div><span aria-current="page" class="font-medium text-foreground overflow-hidden text-ellipsis whitespace-nowrap">Complete Documentation</span>
</nav>
<div id="content" role="main">
<section id="complete-documentation">
<h1>Complete Documentation<a class="headerlink" href="#complete-documentation" title="Link to this heading"><span>#</span></a></h1>
<a class="reference internal image-reference" href="_images/h2m-logo-final.png"><img alt="_images/h2m-logo-final.png" class="align-left" src="_images/h2m-logo-final.png" style="width: 150px;"/></a>
<section id="h2m-output-data-description">
<h2>H2M Output Data Description<a class="headerlink" href="#h2m-output-data-description" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m-output-data-description'"><span>#</span></a></h2>
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>.dataframe tbody tr th {
    vertical-align: middle;
}

.dataframe thead th {
    text-align: middle;
}
</pre></div>
</div>
</style>
<table border="1" class="dataframe">
<thead>
<tr>
<th></th>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>gene_name_h</td>
<td>Human gene name</td>
</tr>
<tr>
<th>1</th>
<td>gene_id_h</td>
<td>Human gene ID</td>
</tr>
<tr>
<th>2</th>
<td>tx_id_h</td>
<td>Human transcript ID</td>
</tr>
<tr>
<th>3</th>
<td>chr_h</td>
<td>Human chromosome number</td>
</tr>
<tr>
<th>4</th>
<td>exon_num_h</td>
<td>Total number of exons of the human transcript</td>
</tr>
<tr>
<th>5</th>
<td>strand_h</td>
<td>Positive or Negative strand of the human transcript on the chromosome</td>
</tr>
<tr>
<th>6</th>
<td>match</td>
<td>The computed reference sequence by given coordinate is matched with the input reference sequence or not</td>
</tr>
<tr>
<th>7</th>
<td>start_h | end_h</td>
<td>Start and end position of the human variant on the chromosome in MAF format</td>
</tr>
<tr>
<th>8</th>
<td>ref_seq_h | alt_seq_h</td>
<td>Reference and alternate sequence of the human variant on the chromosome in MAF format</td>
</tr>
<tr>
<th>9</th>
<td>HGVSc_h | HGVSp_h</td>
<td>HGVSc and HGVSp expression of the human variant</td>
</tr>
<tr>
<th>10</th>
<td>classification_h</td>
<td>Human variant effect classification, including missense/nonsense/in-frame indel/fram-shift indel/intron, etc.</td>
</tr>
<tr>
<th>11</th>
<td>exon_h</td>
<td>Exon/Intron location of the given human mutation, for example, E_7/I_5</td>
</tr>
<tr>
<th>12</th>
<td>type_h</td>
<td>Human variant type in MAF format, including SNP/DNP/TNO/ONP/INS/DEL</td>
</tr>
<tr>
<th>13</th>
<td>status</td>
<td>This mutation can be modeled in the given target transcript or not, True or False</td>
</tr>
<tr>
<th>14</th>
<td>class</td>
<td>H2M modeling result class, 0-5</td>
</tr>
<tr>
<th>15</th>
<td>statement</td>
<td>Statement of the H2M result class</td>
</tr>
<tr>
<th>16</th>
<td>flank_size_left | flank_size_right</td>
<td>Length of the identical sequences between human and mouse on the left/right side of the mutation</td>
</tr>
<tr>
<th>17</th>
<td>gene_name_m</td>
<td>Mouse gene name</td>
</tr>
<tr>
<th>18</th>
<td>gene_id_m</td>
<td>Mouse gene ID</td>
</tr>
<tr>
<th>19</th>
<td>tx_id_m</td>
<td>Mouse transcript ID</td>
</tr>
<tr>
<th>20</th>
<td>chr_m</td>
<td>Mouse chromosome number</td>
</tr>
<tr>
<th>21</th>
<td>exon_num_m</td>
<td>Total number of exons of the mouse transcript</td>
</tr>
<tr>
<th>22</th>
<td>strand_m</td>
<td>Positive or negative strand of the mouse transcript on the chromosome</td>
</tr>
<tr>
<th>23</th>
<td>type_m</td>
<td>Mouse variant type in MAF format, including SNP/DNP/TNO/ONP/INS/DEL</td>
</tr>
<tr>
<th>24</th>
<td>classification_m</td>
<td>Mouse variant effect classification</td>
</tr>
<tr>
<th>25</th>
<td>exon_m</td>
<td>Exon/Intron location of the murine mutation</td>
</tr>
<tr>
<th>26</th>
<td>start_m_ori | end_m_ori</td>
<td>Start and end position of the mouse variant (with exactly the same DNA change) on the chromosome in MAF format</td>
</tr>
<tr>
<th>27</th>
<td>ref_seq_m_ori | alt_seq_m_ori</td>
<td>Reference and alternate sequence of the mouse variant (with exactly the same DNA change) on the chromosome in MAF format</td>
</tr>
<tr>
<th>28</th>
<td>HGVSc_m_ori | HGVSp_m_ori</td>
<td>HGVSc and HGVSp expression of the mouse variant (with exactly the same DNA change)</td>
</tr>
<tr>
<th>29</th>
<td>start_m | end_m</td>
<td>Start and end position of the mouse variant (with the same amino acid change) on the chromosome in MAF format</td>
</tr>
<tr>
<th>30</th>
<td>ref_seq_m | alt_seq_m</td>
<td>Reference and alternate sequence of the mouse variant (with the same amino acid change) on the chromosome in MAF format</td>
</tr>
<tr>
<th>31</th>
<td>HGVSc_m | HGVSp_m</td>
<td>HGVSc and HGVSp expression of the mouse variant (with the same amino acid change)</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="h2m-modeling-class-description">
<h2>H2M Modeling Class Description<a class="headerlink" href="#h2m-modeling-class-description" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m-modeling-class-description'"><span>#</span></a></h2>
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>.dataframe tbody tr th {
    vertical-align: middle;
}

.dataframe thead th {
    text-align: left;
}
</pre></div>
</div>
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>Class</th>
<th>Statement</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>Class 0</td>
<td>This mutation can be originally modeled.</td>
</tr>
<tr>
<th>1</th>
<td>Class 1</td>
<td>This mutation can be alternatively modeled.</td>
</tr>
<tr>
<th>2</th>
<td>Class 2</td>
<td>This mutation can be modeled, but the effect may not be consistent.</td>
</tr>
<tr>
<th>3</th>
<td>Class 3</td>
<td>This mutation cannot be originally modeled and no alternative is found.</td>
</tr>
<tr>
<th>4</th>
<td>Class 4</td>
<td>Mutated sequences are not identical.</td>
</tr>
<tr>
<th>5</th>
<td>Class 5</td>
<td>Coordinate error. This mutation is not in the query gene.</td>
</tr>
<tr>
<th>6</th>
<td>Class 6</td>
<td>This mutation cannot be originally modeled.</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="functions">
<h2>Functions<a class="headerlink" href="#functions" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#functions'"><span>#</span></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.genome_loader">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">genome_loader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.genome_loader" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.genome_loader'"><span>#</span></a></dt>
<dd><p>Load the refernce genome file.</p>
<dl>
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>path (str): path of the genome file.</p></li>
</ul>
</dd>
<dt>Return:</dt><dd><ul class="simple">
<li><p>reference genome records and the index list of chromosomes.</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">records_h</span><span class="p">,</span> <span class="n">index_list_h</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">genome_loader</span><span class="p">(</span><span class="n">path_h_ref</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.anno_loader">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">anno_loader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.anno_loader" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.anno_loader'"><span>#</span></a></dt>
<dd><p>Load the GENCODE annotation file.</p>
<dl>
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>path (str): path of the annotation file.</p></li>
</ul>
</dd>
<dt>Return:</dt><dd><ul class="simple">
<li><p>a FeatureDB</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">db_h</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">anno_loader</span><span class="p">(</span><span class="n">path_h_anno</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.cbio_reader">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">cbio_reader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.cbio_reader" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.cbio_reader'"><span>#</span></a></dt>
<dd><p>Generate MAF-sty h2m input from cbioportal data.</p>
<dl>
<dt>Parameter:</dt><dd><ul class="simple">
<li><p>path (str): the path of mutation data in txt format.</p></li>
<li><p>keep (bool): True: keep all the original columns in the dataframe/ False: keep the necesssary columns for h2m only. Default to False.</p></li>
</ul>
</dd>
<dt>Output: </dt><dd><p>An input dataframe for h2m modeling.</p>
</dd>
<dt>Example:   </dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">cbio_reader</span><span class="p">(</span><span class="s1">'.../data_mutations.txt'</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.clinvar_reader">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">clinvar_reader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_of_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.clinvar_reader" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.clinvar_reader'"><span>#</span></a></dt>
<dd><p>Generate h2m input from ClinVar data.</p>
<dl>
<dt>Parameter:</dt><dd><ul class="simple">
<li><p>path (str): the path of clinvar renference vcf.gz data.</p></li>
<li><p>list_of_ids (list): the list of variation ids. If no value, the function would output all entries in the ClinVar data file.</p></li>
<li><p>keep (bool): True: keep all the original columns in the dataframe/ False: keep the necesssary columns for h2m only. Default to False.</p></li>
</ul>
</dd>
<dt>Output: </dt><dd><p>An input dataframe for h2m modeling.</p>
</dd>
<dt>Example:   </dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filepath</span> <span class="o">=</span> <span class="s1">'.../GrCh37_clinvar_20230923.vcf.gz'</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variation_ids</span> <span class="o">=</span> <span class="p">[</span><span class="mi">925574</span><span class="p">,</span> <span class="mi">925434</span><span class="p">,</span> <span class="mi">926695</span><span class="p">,</span> <span class="mi">925707</span><span class="p">,</span> <span class="mi">325626</span><span class="p">,</span> <span class="mi">1191613</span><span class="p">,</span> <span class="mi">308061</span><span class="p">,</span> <span class="mi">361149</span><span class="p">,</span> <span class="mi">1205375</span><span class="p">,</span> <span class="mi">208043</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">clinvar_reader</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">variation_ids</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.clinvar_to_maf">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">clinvar_to_maf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.clinvar_to_maf" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.clinvar_to_maf'"><span>#</span></a></dt>
<dd><p>Convert clinvar-style input into maf-style input, while keeping other columns intact.</p>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.vcf_reader">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">vcf_reader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.vcf_reader" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.vcf_reader'"><span>#</span></a></dt>
<dd><p>Generate MAF-style h2m input from VCF-format data, including genomAD data.</p>
<dl>
<dt>Parameter:</dt><dd><ul class="simple">
<li><p>path (str): the path of input csv data.</p></li>
<li><p>keep (bool): True: keep all the original columns in the dataframe/ False: keep the necesssary columns for h2m only. Default to False.</p></li>
</ul>
</dd>
<dt>Output: </dt><dd><p>An input dataframe for h2m modeling.</p>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filepath</span> <span class="o">=</span> <span class="s1">'.../gnomAD_v4.0.0_ENSG00000141510_2024_02_07_11_36_03.csv'</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">vcf_reader</span><span class="p">(</span><span class="s1">''</span><span class="p">,</span><span class="s1">'TP53'</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.vcf_to_maf">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">vcf_to_maf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.vcf_to_maf" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.vcf_to_maf'"><span>#</span></a></dt>
<dd><p>Convert vcf-style input into maf-style input, while keeping other columns intact.</p>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.get_variant_type">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">get_variant_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'type_h'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.get_variant_type" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.get_variant_type'"><span>#</span></a></dt>
<dd><p>Generate h2m/cbio-style neucleotide variant type annotations.</p>
<dl>
<dt>Parameter:</dt><dd><ul class="simple">
<li><p>df (str): a dataframe of mutations including columns for both reference and alternate sequences</p></li>
<li><p>ref_col (str): column name for reference sequences.</p></li>
<li><p>alt_col (str): column name for alternate sequences.</p></li>
</ul>
</dd>
<dt>Output: </dt><dd><p>The input dataframe but with a column of vairant type added.</p>
</dd>
<dt>Example:   </dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">get_variant_type</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">'ref_seq_h'</span><span class="p">,</span><span class="s1">'alt_seq_h'</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.get_tx_id">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">get_tx_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ver</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.get_tx_id" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.get_tx_id'"><span>#</span></a></dt>
<dd><p>Query a human or mouse gene for coordinate and information of all its transcripts. Internet needed.</p>
<dl>
<dt>Parameters:  </dt><dd><ul class="simple">
<li><p>id (str):, identification of a human gene. Multiple input forms are accepted, including gene name, stable ensembl gene id with or without version number.</p></li>
<li><p>species (str): ‘h’ for human or ‘m’ for mouse.</p></li>
<li><p>ver (int): specify the version of human, one of 37/38. It is a necessary parameter.</p></li>
<li><p>ty (str): OPTIONAL. type of your input id. string, one of ‘name’/’gene_id’.</p></li>
<li><p>show (bool): OPTIONAL. print summary of output or not.</p></li>
</ul>
</dd>
<dt>Return:  </dt><dd><ul class="simple">
<li><p>A list [chromosome, start location(of gene), end location(of gene), canonical transcript id, list of all transcript id (the canonical one included and always at the first place), a list of additional information of each transcript]</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'TP53'</span><span class="p">,</span><span class="s1">'h'</span><span class="p">,</span><span class="n">ver</span><span class="o">=</span><span class="mi">37</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.get_tx_batch">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">get_tx_batch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ver</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.get_tx_batch" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.get_tx_batch'"><span>#</span></a></dt>
<dd><p>Batch query of canonical transcript IDs of human or mouse genes.</p>
<dl>
<dt>Parameters:  </dt><dd><ul class="simple">
<li><p>df (Pandas DataFrame): Must include a column of gene names named ‘gene_name_h’/’gene_name_m’, depending on the species. An index column is recommended.</p></li>
<li><p>species (str): ‘h’ for human or ‘m’ for mouse.</p></li>
<li><p>ver (int): specify the version of human, one of 37/38.</p></li>
</ul>
</dd>
<dt>Return:  </dt><dd><ul class="simple">
<li><p>Two dataframes. The first dataframe is the processed original dataframe with canonical transcirpt id attached in the column named ‘tx_id_h’/’tx_id_m’. The second dataframe contains all rows that are not successfully processed.</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_batch</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="s1">'h'</span><span class="p">,</span><span class="n">ver</span><span class="o">=</span><span class="mi">37</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.query">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h2m'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.query" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.query'"><span>#</span></a></dt>
<dd><p>Query homologous mouse genes of human genes.</p>
<dl>
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>id (str): name/gene_id/tx_id of human.</p></li>
<li><p>direction (str): OPTIONAL. query from human gene to the mouse gene (‘h2m’) or vise versa (‘m2h’).</p></li>
<li><p>db (FeatureDB): OPITONAL. The transcript annotation database of specific version.</p></li>
<li><p>ty (str): OPITONAL. Specify the id type. one of ‘gene_id’/’tx_id’/’name’.</p></li>
</ul>
</dd>
<dt>Return:</dt><dd><ul class="simple">
<li><p>a list of human gene name, mouse gene name, mapping type and sequence similarity.</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'TP53'</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.query_batch">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">query_batch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h2m'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.query_batch" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.query_batch'"><span>#</span></a></dt>
<dd><p>Batch query of orthologous mouse gene of given human genes.</p>
<dl>
<dt>Parameters:  </dt><dd><ul class="simple">
<li><p>df (Pandas DataFrame): Must include a column of gene names named ‘gene_name_h’. An index column is recommended.</p></li>
<li><p>direction: OPTIONAL. query from human gene to the mouse gene (‘h2m’) or vise versa (‘m2h’).</p></li>
</ul>
</dd>
<dt>Return:  </dt><dd><ul class="simple">
<li><p>Two dataframes. The first dataframe is the processed original dataframe with canonical transcirpt id attached in the column named ‘gene_name_m’. The second dataframe contains all rows that are not successfully processed.</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">query_batch</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.model">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_list_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">records_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_list_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tx_id_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tx_id_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ty_h</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ver</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h2m'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">search_alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nonstop_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splicing_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_input</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory_protect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.model" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.model'"><span>#</span></a></dt>
<dd><p>Model human variants in the mouse genome.</p>
<dl>
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>records_h, index_list_h, records_m, index_list_m: human and mouse reference genome.</p></li>
<li><p>db_h, db_m: human and mouse GENCODE annotation.</p></li>
<li><p>tx_id_h, tx_id_m: human and mouse transcript id (could get by <strong>h2m.get_tx_id()</strong>). Transcript ids of input and output variants if use direction = ‘h2h’ or  direction = ‘m2m’.</p></li>
<li><p>start_h, end_h: int, start and end location of the mutation on the chromosome.</p></li>
<li><p>ref_seq_h: str, human mutation reference sequence. Reference sequence of the input variant if use direction = ‘m2h’/’h2h’/’m2m’.</p></li>
<li><p>alt_seq_h: str, human mutation alternate sequence. Alternate sequence of the input variant if use direction = ‘m2h’/’h2h’/’m2m’.</p></li>
<li><p>ty_h: str, human variantion type in MAF format. One of [‘SNP’, ‘DNP’, ‘TNP’,’ONP’, ‘INS’, ‘DEL’].</p></li>
<li><p>ver: int, human ref genome number. 37 or 38.</p></li>
<li><p>direction (optional): str, set the modeling direction by ‘h2m’ (default) or ‘m2h’, ‘h2h’, ‘m2m’.</p></li>
<li><p>param (optional): set param = ‘BE’ and will only output base editing modelable results.</p></li>
<li><p>coor (optional): default = ‘nc’. set input = ‘aa’ and will be compatable with input of amino acid variants.</p></li>
<li><p>search_alternative (optional): set search_alternative = False and will only output original modeling results.</p></li>
<li><p>max_alternative (optional): the maximum number of output alternatives of one human variants.</p></li>
<li><p>nonstop_size (optional): the length of neucleotides that are included after the stop codon for alignment and translation in case of the nonstop mutations or frame shifting mutations.</p></li>
<li><p>splicing_size (optional): the number of amino acids or neucleotides (for non-coding mutations) that are included after the top codon for the consideration of frame-shifting effect.</p></li>
<li><p>batch (optional): set batch = True and will use input align_dict to save time in batch processing.</p></li>
<li><p>show_sequence (optional): set batch = True and will output the whole sequences.</p></li>
<li><p>align_dict (optional): input a prepared dictionary of alignment indexes to save time in batch processing.</p></li>
<li><p>memory_protect (optional): default True. Break long alignments that may lead to death of the kernel.</p></li>
<li><p>memory_size (optional): maxlength of aligned sequence when memory_protect == True.</p></li>
</ul>
</dd>
<dt>Other rules:</dt><dd><p>1. If the mutation falls in the coding and non-coding regions at the same time, it would be considered and processed as a ORIGIAL-MODELING ONLY mutation.
3. The alt_seq input should be in the positive strand and the start_h coordinate should be smaller than or equal the end_h coordinate.
4. If the ref-seq or alt-see has no length, it could be input as ‘’ or ‘-‘.</p>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span> <span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.model_batch">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">model_batch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">records_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_list_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">records_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_list_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_h</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ver</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h2m'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">search_alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nonstop_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splicing_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_input</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory_protect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.model_batch" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.model_batch'"><span>#</span></a></dt>
<dd><p>Batch modeling of human variants in the mouse genome.</p>
<dl>
<dt>Parameters:  </dt><dd><ul class="simple">
<li><p>df (Pandas DataFrame): Must include columns {‘start_h’,’end_h’,’type_h’,’ref_seq_h’,’alt_seq_h’,’tx_id_h’,’tx_id_m’,’index’}.</p></li>
<li><p>ecords_h, index_list_h, records_m, index_list_m: reference genome files</p></li>
<li><p>db_h, db_m: genomic annotation files</p></li>
<li><p>ver (int): specify the version of human, one of 37/38.</p></li>
<li><p>param (optional): set param = ‘BE’ and will only output base editing modelable results.</p></li>
<li><p>direction (optional): str, set the modeling direction by ‘h2m’ (default) or ‘m2h’.</p></li>
<li><p>coor (optional): default = ‘nc’. set input = ‘aa’ and will be compatable with input of amino acid variants.</p></li>
<li><p>search_alternative (optional): set search_alternative = False and will only output original modeling results.</p></li>
<li><p>max_alternative (optional): the maximum number of output alternatives of one human variants.</p></li>
<li><p>nonstop_size (optional): the length of neucleotides that are included after the stop codon for alignment and translation in case of the nonstop mutations or frame shifting mutations.</p></li>
<li><p>splicing_size (optional):</p></li>
<li><p>batch (optional): set batch = True and will use input align_dict to save time in batch processing.</p></li>
<li><p>show_sequence (optional): set batch = True and will output the whole sequences.</p></li>
<li><p>align_dict (optional): input a prepared dictionary of alignment indexes to save time in batch processing.</p></li>
<li><p>memory_protect (optional): default True. Break long alignments that may lead to death of the kernel.</p></li>
<li><p>memory_size (optional): maxlength of aligned sequence when memory_protect == True.</p></li>
<li><p>bind (optional): to bind the output dataframe with the original input or not.</p></li>
</ul>
</dd>
<dt>Return:  </dt><dd><ul class="simple">
<li><p>Two dataframes. The first dataframe is the processed original dataframe. The second dataframe contains all rows that are not successfully processed.</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">model_batch</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">records_h</span><span class="p">,</span> <span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> <span class="n">param</span> <span class="o">=</span> <span class="s1">'BE'</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
<dl class="py function">
<dt class="sig sig-object py" id="h2m.visualization">
<span class="sig-prename descclassname"><span class="pre">h2m.</span></span><span class="sig-name descname"><span class="pre">visualization</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_result</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#h2m.visualization" title="Link to this definition" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#h2m.visualization'"><span>#</span></a></dt>
<dd><p>Visualize h2m modeling results.</p>
<dl>
<dt>Parameter:</dt><dd><ul class="simple">
<li><p>model_result (list): the output of <cite>h2m.model(show_sequence = True)</cite> function.</p></li>
<li><p>flank_size (int) (de).</p></li>
<li><p>print_size (int): lenth of neucleotide/peptide included on both sides of the flank region.</p></li>
</ul>
</dd>
<dt>Output: </dt><dd><p>A visualization plot.</p>
</dd>
<dt>Example:   </dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span><span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> <span class="n">show_sequence</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">h2m</span><span class="o">.</span><span class="n">visualization</span><span class="p">(</span><span class="n">model_result</span><span class="p">,</span> <span class="n">flank_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">print_size</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>
</section>
</section>
</div></div><aside class="hidden text-sm xl:block" id="right-sidebar">
<div class="sticky top-16 -mt-10 max-h-[calc(100vh-5rem)] overflow-y-auto pt-6 space-y-2"><p class="font-medium">On this page</p>
<ul>
<li><a :data-current="activeSection === '#h2m-output-data-description'" class="reference internal" href="#h2m-output-data-description">H2M Output Data Description</a></li>
<li><a :data-current="activeSection === '#h2m-modeling-class-description'" class="reference internal" href="#h2m-modeling-class-description">H2M Modeling Class Description</a></li>
<li><a :data-current="activeSection === '#functions'" class="reference internal" href="#functions">Functions</a><ul>
<li><a :data-current="activeSection === '#h2m.genome_loader'" class="reference internal" href="#h2m.genome_loader"><code class="docutils literal notranslate"><span class="pre">genome_loader()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.anno_loader'" class="reference internal" href="#h2m.anno_loader"><code class="docutils literal notranslate"><span class="pre">anno_loader()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.cbio_reader'" class="reference internal" href="#h2m.cbio_reader"><code class="docutils literal notranslate"><span class="pre">cbio_reader()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.clinvar_reader'" class="reference internal" href="#h2m.clinvar_reader"><code class="docutils literal notranslate"><span class="pre">clinvar_reader()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.clinvar_to_maf'" class="reference internal" href="#h2m.clinvar_to_maf"><code class="docutils literal notranslate"><span class="pre">clinvar_to_maf()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.vcf_reader'" class="reference internal" href="#h2m.vcf_reader"><code class="docutils literal notranslate"><span class="pre">vcf_reader()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.vcf_to_maf'" class="reference internal" href="#h2m.vcf_to_maf"><code class="docutils literal notranslate"><span class="pre">vcf_to_maf()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.get_variant_type'" class="reference internal" href="#h2m.get_variant_type"><code class="docutils literal notranslate"><span class="pre">get_variant_type()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.get_tx_id'" class="reference internal" href="#h2m.get_tx_id"><code class="docutils literal notranslate"><span class="pre">get_tx_id()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.get_tx_batch'" class="reference internal" href="#h2m.get_tx_batch"><code class="docutils literal notranslate"><span class="pre">get_tx_batch()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.query'" class="reference internal" href="#h2m.query"><code class="docutils literal notranslate"><span class="pre">query()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.query_batch'" class="reference internal" href="#h2m.query_batch"><code class="docutils literal notranslate"><span class="pre">query_batch()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.model'" class="reference internal" href="#h2m.model"><code class="docutils literal notranslate"><span class="pre">model()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.model_batch'" class="reference internal" href="#h2m.model_batch"><code class="docutils literal notranslate"><span class="pre">model_batch()</span></code></a></li>
<li><a :data-current="activeSection === '#h2m.visualization'" class="reference internal" href="#h2m.visualization"><code class="docutils literal notranslate"><span class="pre">visualization()</span></code></a></li>
</ul>
</li>
</ul>
</div>
</aside>
</main>
</div>
</div><footer class="py-6 border-t border-border md:py-0">
<div class="container flex flex-col items-center justify-between gap-4 md:h-24 md:flex-row">
<div class="flex flex-col items-center gap-4 px-8 md:flex-row md:gap-2 md:px-0">
<p class="text-sm leading-loose text-center text-muted-foreground md:text-left">© 2024, Kexin Dong Built with <a class="font-medium underline underline-offset-4" href="https://www.sphinx-doc.org" rel="noreferrer">Sphinx 7.2.6</a></p>
</div>
</div>
</footer>
</div>
<script src="_static/documentation_options.js?v=8d563738"></script>
<script src="_static/doctools.js?v=888ff710"></script>
<script src="_static/sphinx_highlight.js?v=dc90522c"></script>
<script defer="defer" src="_static/theme.js?v=8dc3ddfe"></script>
</body>
</html>