Index: wordcloud_func.py
===================================================================
diff --git a/wordcloud_func.py b/wordcloud_func.py
deleted file mode 100644
--- a/wordcloud_func.py	(revision 7237d18beebcbb857f68c3c304231f98ab8d03b9)
+++ /dev/null	(revision 7237d18beebcbb857f68c3c304231f98ab8d03b9)
@@ -1,36 +0,0 @@
-"""
-Function to create and write an eliptically shaped wordcloud in the Susmon colour palette
-from a list of word frequencies.
-"""
-from wordcloud import WordCloud
-import numpy as np
-import random
-from PIL import Image, ImageDraw
-
-
-# funcion to return a colour from the Susmon colour palette
-def _susmon_colour_func(word, font_size, position, orientation, random_state=None, **kwargs):
-    susmon_colours = [(109, 110, 113), (240, 80, 70), (150, 80, 150), (226, 228, 30),
-                      (141, 198, 63), (20, 190, 240), (160, 228, 247), (35, 130, 202),
-                      (0, 180, 170), (253, 185, 52), (65, 64, 66)]
-    return f"rgb{random.choice(susmon_colours)}"
-
-
-def create_and_write_wordcloud(frequencies, wcloud_file):
-    # create an elliptical mask for our wordcloud
-    w, h = 600, 500
-    shape = [(20, 20), (w - 30, h - 150)]
-    im = Image.new(mode="RGB", size=(w, h), color='white')
-    im2 = ImageDraw.Draw(im)
-    im2.ellipse(shape, fill='black', outline='black')
-    mask = np.array(im)
-
-    # Create the wordcloud object and write to file
-    try:
-        wordcloud = WordCloud(font_path="Arial.ttf", background_color="white", color_func=_susmon_colour_func,
-                              mask=mask, width=560, height=560, margin=0, max_words=25,
-                              prefer_horizontal=1).generate_from_frequencies(frequencies)
-
-        wordcloud.to_file(wcloud_file)
-    except:
-        print(f"Unable to create wordcloud: {wcloud_file}")
